File: simulations-java/common/scala-common/scala-src/edu/colorado/phet/scalacommon/record/PlaybackSpeedSlider.scala
package edu.colorado.phet.scalacommon.record

import _root_.edu.colorado.phet.common.piccolophet.event.CursorHandler
import common.phetcommon.resources.PhetCommonResources
import edu.colorado.phet.common.phetcommon.math.Function.LinearFunction
import edu.colorado.phet.common.phetcommon.view.controls.valuecontrol.LinearValueControl
import java.awt.Color
import java.awt.Cursor
import java.util.Hashtable
import javax.swing.event.{ChangeListener, ChangeEvent}
import javax.swing.{JSlider, JLabel}

import umd.cs.piccolo.PNode
import umd.cs.piccolox.pswing.PSwing
import common.phetcommon.resources.PhetCommonResources._

class PlaybackSpeedSlider[T](model: RecordModel[T]) extends PNode {
  addInputEventListener(new CursorHandler)
  val slider = new JSlider
  slider.setBackground(new Color(0, 0, 0, 0))
  val transform = new LinearFunction(slider.getMinimum, slider.getMaximum, 0.25, 2.0)

  val dict = new Hashtable[Integer, JLabel]

  dict.put(slider.getMinimum, new JLabel(getString("Common.time.slow")))
  dict.put(slider.getMaximum, new JLabel(getString("Common.time.fast")))

  slider.setLabelTable(dict)
  slider.setPaintLabels(true)
  val playbackSpeedSlider = new PSwing(slider)
  addChild(playbackSpeedSlider)
  slider.addChangeListener(new ChangeListener() {
    def stateChanged(e: ChangeEvent) = {
      model.setPlayback(transform.evaluate(slider.getValue))
    }
  })

  def updatePlaybackSliderVisible = playbackSpeedSlider.setVisible(model.isPlayback)
  model.addListener(() => {updatePlaybackSliderVisible})
  updatePlaybackSliderVisible
}

File: trunk/simulations-java/common/record-and-playback/src/edu/colorado/phet/recordandplayback/gui/PlaybackSpeedSlider.java
package edu.colorado.phet.recordandplayback.gui;

import edu.colorado.phet.common.phetcommon.math.Function;
import edu.colorado.phet.common.phetcommon.resources.PhetCommonResources;
import edu.colorado.phet.common.phetcommon.util.SimpleObserver;
import edu.colorado.phet.common.piccolophet.event.CursorHandler;
import edu.colorado.phet.recordandplayback.model.RecordAndPlaybackModel;
import edu.umd.cs.piccolo.PNode;
import edu.umd.cs.piccolox.pswing.PSwing;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.util.Hashtable;

/**
 * The PlaybackSpeedSlider allows the user to view and modify the playback speed for a recorded sequence.
 *
 * @author Sam Reid
 * @param <T>
 */
public class PlaybackSpeedSlider<T> extends PNode {
    private RecordAndPlaybackModel<T> model;

    public PlaybackSpeedSlider(final RecordAndPlaybackModel<T> model) {
        this.model = model;
        addInputEventListener(new CursorHandler());

        final JSlider slider = new JSlider();
        slider.setBackground(new Color(0, 0, 0, 0));
        final Function.LinearFunction transform = new Function.LinearFunction(slider.getMinimum(), slider.getMaximum(), 0.25, 2.0);

        Hashtable<Integer, JLabel> dict = new Hashtable<Integer, JLabel>();

        dict.put(slider.getMinimum(), new JLabel(PhetCommonResources.getString("Common.time.slow")));
        dict.put(slider.getMaximum(), new JLabel(PhetCommonResources.getString("Common.time.fast")));

        slider.setLabelTable(dict);
        slider.setPaintLabels(true);
        PSwing playbackSpeedSlider = new PSwing(slider);
        addChild(playbackSpeedSlider);
        slider.addChangeListener(new ChangeListener() {
            public void stateChanged(ChangeEvent e) {
                model.setPlayback(transform.evaluate(slider.getValue()));
            }
        });

        SimpleObserver updatePlaybackSliderVisible = new SimpleObserver() {
            public void update() {
                setVisible(model.isPlayback());
            }
        };
        model.addObserver(updatePlaybackSliderVisible);
        updatePlaybackSliderVisible.update();
    }
}